python/singa/model.py

```diff
@@ -0,0 +1,238 @@
+        self.sequential = sequential
+
+    def forward(self, *input):
+        """Defines the computation performed at every call.
+
+        Should be overridden by all subclasses.
+
+        Args:
+            *input: the input training data for the model
+
+        Returns:
+            out: the outputs of the forward propagation.
+        """
+        raise NotImplementedError
+
+    def train_one_batch(self, *input):
+        raise NotImplementedError
+
+    def train(self, mode=True):
+        """Set the model in evaluation mode.
+
+        Args:
+            mode(bool): when mode is True, this model will enter training mode
+        """
+        self.training = mode
+        autograd.training = mode
+
+    def eval(self):
+        """Sets the model in evaluation mode.
+        """
+        self.train(mode=False)
+        autograd.training = False
+
+    def graph(self, mode=True, sequential=False):
+        """ Turn on the computational graph. Specify execution mode.
+
+        Args:
+            mode(bool): when mode is True, model will use computational graph
+            sequential(bool): when sequential is True, model will execute ops
+            in the graph follow the order of joining the graph
+        """
+        self.graph_mode = mode
+        self.sequential = sequential
+
+    def on_device(self, device):
+        """Sets the target device.
+
+        The following training will be performed on that device.
+
+        Args:
+            device(Device): the target device
+        """
+        self._device = device
+
+    def __get_name__(self):
+        return self.__class__.__name__
+
+    def __call__(self, *input, **kwargs):
+        if self.training:
+            return self.train_one_batch(*input, **kwargs)
+        else:
+            return self.forward(*input, **kwargs)
+
+    def _flatten_dictionary(self,d, parent_key='', sep=':'):
+        items = []
+        for k, v in d.items():
+            new_key = parent_key + sep + k if parent_key else k
+            if isinstance(v, dict):
+                items.extend(self._flatten_dictionary(v, new_key, sep=sep).items())
+            else:
+                items.append((new_key, v))
+        return dict(items)
+
+    def _de_flatten_dictionary(self,d):
+        def extend_d_tree(leaf, k, v, sep=":"):
+            if sep in k:
+                lead_k, trail_k = k.split(sep,1)
+                new_leaf = leaf.setdefault(lead_k, {})
+                extend_d_tree(new_leaf, trail_k, v)
+            else:
+                leaf[k] = v
+
+        root=dict()
+        for k, v in d.items():
+            extend_d_tree(root, k, v)
+        return root
+
+    def get_params(self):
+        params = super().get_params()
+        # input params are nested {"l1": {"l2": {"l3": {"W": []}}}}
+        # flatten into {"l1:l2:l3:W": []}
+        # recursively flatten params nested dict
+        return self._flatten_dictionary(params)
+
+    def set_params(self, params):
```

Comment:
if params is a flatten dict, we can implement simply like

1. for each sublayer of this layer
       sublayer.set_params(params)

2. for each param of this layer
      param.copy_from(params[param.name])